### PR TITLE
[SYCL][Doc] Add SYCL marray complex extension proposal

### PR DESCRIPTION
**Pull request by AidanBeltonS**
This PR adds an extension that proposes specializing the `marray` class to support the SYCL complex extension in #6233. The `marray` class is slightly modified, with some inapplicable operators removed for the specialization. No new functions or operators are added to the `marray` class.

Five overloaded free functions are added to provide complex functionality to the `marray` specialization; `make_complex_marray`, `get_real`/`get_imag`, and `set_real`/`set_imag`. They provide the functionality normally given by constructors and member functions `real` and `imag`.

Finally, all relevant complex math functions are extended to support this class.

### CODE REVIEW

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+marrays and adding new free functions to simplify accessing, setting marray
+values, and construct complex marrays.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_MARRAY_COMPLEX` to one of the values defined in the table
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:72]: Little confusing with `SYCL_EXT_ONEAPI_MARRAY_COMPLEX` as the feature test macro in `sycl_ext_oneapi_complex_marray`.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 2]
```diff
@@ -0,0 +1,543 @@
+= sycl_ext_oneapi_complex
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:1]: Name should match filename.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 5 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:41]: The dependence on the complex extension should also be mentioned here, with links preferably.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+            friend marray<bool, NumElements> operator!(const marray &v) = delete;
+    }
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
+
+The list of deleted operators are: %, %=, ++, --, +, -, &, |, ^, &=, |=, ^=,
+&&, ||, <<, >>, <<=, >>=, <, >, <=, >=, ~, !
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:231]: ASCIIDOC seems to want to convert `<=` to an arrow. Must be a way to prevent that.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+
+  template <class T, std::size_t NumElements>
+  marray<T, NumElements> get_imag(const marray<sycl::ext::oneapi::complex<T>, NumElements> &input);
+
+  // return element of component
+  template <class T, std::size_t NumElements>
+  T get_real(const marray<sycl::ext::oneapi::complex<T>, NumElements> &input, std::size_t index);
+
+  template <class T, std::size_t NumElements>
+  T get_imag(const marray<sycl::ext::oneapi::complex<T>, NumElements> &input, std::size_t index);
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:276]: What are the values in having these two free functions? Could a user not just get the element using `input[index]` and call `imag`/`real` on that?

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+
+  // return sequence of elements of component
+  template <class T, std::size_t NumElements, std::size_t... I>
+  T get_real(const marray<sycl::ext::oneapi::complex<T>, NumElements> &input, std::integer_sequence<std::size_t, I...> int_seq);
+
+  template <class T, std::size_t NumElements, std::size_t... I>
+  T get_imag(const marray<sycl::ext::oneapi::complex<T>, NumElements> &input, std::integer_sequence<std::size_t, I...> int_seq);
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:283]: Is the return type of this correct?

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+namespace oneapi {
+
+  template <std::size_t NumElements>
+  class marray<double, NumElements>;
+
+  template <std::size_t NumElements>
+  class marray<float, NumElements>;
+
+  template <std::size_t NumElements>
+  class marray<sycl::half, NumElements>;
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:325]: What are these for? Are they missing `complex`?

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+
+  template <class T, std::size_t NumElements>
+  void set_imag(marray<sycl::ext::oneapi::complex<T>, NumElements> &input, std::size_t index, const T value);
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
+
+The class `sycl::oneapi::marray<complex<T>, N>`, has specializations
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:310]:
```suggestion
The class `sycl::ext::oneapi::marray<sycl::ext::oneapi::complex<T>, N>`, has specializations
```
Right?

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+  class marray<sycl::half, NumElements>;
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
+
+The generic type `mgencomplex` is defined as types
+`marray<complex<double>, {N}>`, `complex<complex<float>, {N}>`,
+`complex<complex<sycl::half>, {N}>`.
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:334]:
```suggestion
`marray<complex<double>, {N}>`, `marray<complex<float>, {N}>`,
`marray<complex<sycl::half>, {N}>`.
```

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+|Set each element of the imaginary components in x to the corresponding element in y.
+|`void set_real(mgencomplex& x, const genfloat& y);`
+|Set each element of the real components in x to the decimal number y.
+|`void set_imag(mgencomplex& x, const genfloat& y);`
+|Set each element of the imaginary components in x to the decimal number y.
+|`void set_real(mgencomplex& x, std::size_t idx, const genfloat& y);`
+|Set the real value of element idx in x to the decimal number y.
+|`void set_imag(mgencomplex& x, std::size_t idx, const genfloat& y);`
+|Set the imaginary value of element idx in x to the decimal number y.
+
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:379]:
```suggestion
|===

```

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 4 above)
Comment (gmlueck) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:231]: Enclose with `+` like:

```
+<=+
```

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 1 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: Agreed I have changed the feature test macro to `SYCL_EXT_ONEAPI_COMPLEX_MARRAY`

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 2 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: Done

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 3 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:41]: I have added a bit about the dependence along with a link to the complex extension.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 4 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: Done

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 5 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:278]: These two functions are for completeness, it would function identically to `arr[index].real()`.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 6 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: The return type is wrong. I have updated it to return an marray of the same size as the integer sequence.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 5 above)
Comment (steffenlarsen) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:278]: Personally I think I would prefer to be without it. I can see the usefulness of the others though. @gmlueck - Thoughts?

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 7 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: This was a typo, these should be specialized with `complex<sycl::half>`, etc. I have updated it.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 8 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: Done

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 9 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: Done

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 10 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: Done

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+            friend marray<bool, NumElements> operator!(const marray &v) = delete;
+    }
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
+
+The list of deleted operators are: %, %=, ++, --, +, -, &, |, ^, &=, |=, ^=,
+&&, ||, <<, >>, <<=, >>=, <, >, +<=+, >=, ~, !
```
Comment (gmlueck) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:233]: I think there is no need to create a specialization of `marray` for `sycl::ext::oneapi::complex` because the standard SYCL definition should be sufficient.  Note that the standard SYCL `marray` only requires that the underlying type `T` is a C++ [NumericType](https://en.cppreference.com/w/cpp/named_req/NumericType), which is quite expansive.

It looks to me like `sycl::ext::oneapi::complex` does not currently meet the requirements of NumericType because it is missing a default constructor and is missing an assignment operator.  However, these seem like oversights that should be fixed.

You will probably ask about the `marray` operators that make no sense for complex types.  This, I think, is a bug in the SYCL spec that should be fixed.  Essentially, the spec should say that each operator is only available if the underlying type `T` implements that operator.  See https://gitlab.khronos.org/sycl/Specification/-/issues/559 (requires Khronos access).

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 3 above)
Comment (gmlueck) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:41]: Just a nit, but it would be better to note the dependency in the "Dependencies" section, rather than down below in the "Specification" section.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 5 above)
Comment (gmlueck) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:278]: I think I would leave them out too.  They are more verbose than the existing alternative, so it's not clear what benefit they provide.  Note this applies also to the "set" versions.  For example, consider what they will look like once they get adopted into the `sycl` namespace:

```
float f = m[5].real();
float f = sycl::get_real(m, 5);  // Still more verbose

m[5].real(3.14);
sycl::set_real(m, 5, 3.14);  // Again more verbose
```

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 11 above)
Comment (gmlueck) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:233]: These operators are still not rendered correctly in HTML, but given my previous comment, I think these lines will just be deleted anyways.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 5 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: I don't have any opposition to removing if it feels redundant, so I have removed it.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 3 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:41]: I have moved the text up to the "Dependencies" section

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 11 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: Okay, I have updated this proposal to use the different `marray` definition. Instead this proposal just instantiates the `marray` class for the various `gencomplex` types.

Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: I have removed the operator list due to the above comment.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+proposed.
+
+[%header,cols="5,5"]
+|===
+|Function
+|Description
+
+|`mgencomplex make_complex_marray(const mgenfloat& x, const mgenfloat& y);`
+|Constructs a marray of complex numbers with real values in marray x, and the imaginary values in marray y.
+|`mgencomplex make_complex_marray(const mgenfloat& x, const genfloat& y);`
```
Comment (gmlueck) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:211]: Just checking ... is it your intention to pass this scalar value `y` by reference rather than by value?

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
+
+The generic type `mgencomplex` is defined as types
+`marray<sycl::ext::oneapi::complex<double>, {N}>`,
+`marray<sycl::ext::oneapi::complex<float>, {N}>`,
+`marray<sycl::ext::oneapi::complex<sycl::half>, {N}>`.
```
Comment (gmlueck) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:198]: I think we should make it clear that the extension does not define a C++ type named `mgencomplex`.  Maybe a better way to say it is:

> The table below shows free functions operating on the `marray` type when it is specialized with `sycl::ext::oneapi::complex<double>`, `sycl::ext::oneapi::complex<float>`, and `sycl::ext::oneapi::complex<sycl::half>`.  For the purposes of this specification, we use the generic type name `mgencomplex` to represent these three specializations.  However, there is no C++ type actually named `mgencomplex`.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+  template <class T, std::size_t NumElements>
+  void set_imag(marray<sycl::ext::oneapi::complex<T>, NumElements> &input, const T value);
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
+
+The class `sycl::ext::oneapi::marray<sycl::ext::oneapi::complex<T>, N>`,
+has instantiations of `T`; `float`, `double`, and `sycl::half` declared.
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: Okay I have removed this bit and the three specializations.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 12 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: That is a good catch. I have removed `genfloat` from being passed by reference to this and the `set_X` free functions.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 13 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: I have clarified the paragraph, making it clear that there is no C++ type named `mgencomplex`

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+namespace ext {
+namespace oneapi {
+
+  // Make_complex_marray
+
+  template <class T, std::size_t NumElements>
+  marray<sycl::ext::oneapi::complex<T>, NumElements> make_complex_marray(const marray<T, NumElements> &real, const marray<T, NumElements> &imag);
+
+  template <class T, std::size_t NumElements>
+  marray<sycl::ext::oneapi::complex<T>, NumElements> make_complex_marray(const marray<T, NumElements> &real, const T imag);
```
Comment (gmlueck) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:127]:
```suggestion
  marray<sycl::ext::oneapi::complex<T>, NumElements> make_complex_marray(const marray<T, NumElements> &real, T imag);
```

`const` is not meaningful when applied to a parameter that is passed by value.  Same for other parameters below that were changed to pass-by-value.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+
+    mgencomplex log(const mgencomplex& x);
+
+    mgencomplex log10(const mgencomplex& x);
+
+    mgenfloat norm(const mgencomplex& x);
+
+    mgencomplex polar(const mgenfloat& rho, const mgenfloat& theta);
+    mgencomplex polar(const mgenfloat& rho, const genfloat& theta = 0);
+    mgencomplex polar(const genfloat& rho, const mgenfloat& theta);
```
Comment (gmlueck) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:271]:
```suggestion
    mgencomplex polar(const mgenfloat& rho, genfloat theta = 0);
    mgencomplex polar(genfloat rho, const mgenfloat& theta);
```

Should these (and other) mathematical operations be changed to pass-by-value?

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 15 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: Done

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 16 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: Yes pow and polar have these overloads changed to pass-by-value

**Review by gmlueck — APPROVED** [review]

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== Marray Complex Class Specialization
```
Comment (rolandschulz) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:92]: This is confusing wording given that on line 101 you say "As the `marray` class is not specialized in this proposal."

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+
+The `marray` class is extended for the `sycl::ext::oneapi::complex` class.
```
Comment (rolandschulz) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:94]: Why is that needed? SYCL 2022 says "The element type parameter, DataT, must be a numeric type as it is defined by C++ standard.". Therefore marray already supports complex. I think everything up line 106 isn't needed.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+`marray` interface. The usage of free functions does cause a deviation
+from the `std::complex` interface. However, it does reduce this extensions
+impact on the `marray` interface.
+
+```C++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+  // Make_complex_marray
```
Comment (rolandschulz) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:121]: Should we really add multiple `make_complex_marray` overloads just for complex? Wouldn't it make more sense to add a general purpose generator expression mechanism? E.g. like the one of std::simd in parallelism TSv2.
If we add `template<class G> marray(G&& gen)` instead of writing `make_complex_marray(ar, ai)` (where ar, ai are marrays) one could write `marray([&](auto i){return complex{ar[i], ai[i]};}`. Isn't that simple enough?

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 20]
```diff
@@ -0,0 +1,379 @@
...
+  marray<sycl::ext::oneapi::complex<T>, NumElements> make_complex_marray(const marray<T, NumElements> &real, const marray<T, NumElements> &imag, std::integer_sequence<std::size_t, I...> int_seq);
+
+  template <class T, std::size_t NumElements, std::size_t... I>
+  marray<sycl::ext::oneapi::complex<T>, NumElements> make_complex_marray(const marray<sycl::ext::oneapi::complex<T>, NumElements> &cmplx, std::integer_sequence<std::size_t, I...> int_seq);
+
+  // Get
+
+  // return marray of component
+  template <class T, std::size_t NumElements>
+  marray<T, NumElements> get_real(const marray<sycl::ext::oneapi::complex<T>, NumElements> &input);
```
Comment (rolandschulz) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:142]: Same as for creating, do we need this? If we can't instead add e.g. map/transform, then `get_real(a)` would become `map(a, [](auto e){return e.real();}`.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 17 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: Agreed I have updated it to say instantiation.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 18 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:94]: So I am not sure that `complex` is a numeric type. So far as I understand (please correct me if I am wrong), a type is numeric if it has `std::numeric_limits` defined for it. `std::complex` does not appear to have a numeric limits specialization. While `std::numeric_limits` could be specialized for complex, a lot of the constants and member functions don't make sense. For example, what is the complex equivalent of infinity, is it `(inf, inf)` does that mean `(inf, 0)` is not infinity although its magnitude is infinity?

I think the safer and simpler way of viewing `complex` is as two numeric types, but it itself is not a numeric type.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 19 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:121]: Applications using arrays of complex values will likely want an interface that performs similar operations to the `std::complex` interface. I.e. being able to construct a complex value with a real and imaginary vector component. While `marray([&](auto i){return complex{ar[i], ai[i]};}` would work, it is much more verbose than `make_complex_marray(ar, ai)`, for something that is a simple element-wise constructor and it would be a significant departure from the simple `std::complex` constructors.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 20 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:142]: My opinion is the same as above. This is more verbose and when dealing with vector math it would look much more messy.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 18 above)
Comment (gmlueck) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:94]: Here is the cppreference definition of `NumericType`:

https://en.cppreference.com/w/cpp/named_req/NumericType

It does not say that `std::numeric_limits` must be defined.

My feedback on these lines through 106:

* I agree with @rolandschulz that the line "The `marray` class is extended for the `sycl::ext::oneapi::complex` class." is unneeded.  I do not think we are extending that class.  In fact, the next sentence says that the interface is not changed, so these two statements seems to be contradictory.

* The statement that "The `marray` complex instantiation is trivially copyable" seems useful.  I don't think we currently have anything in the SYCL 2020 spec that guarantees this, though there is an open Khronos issue about it (https://gitlab.khronos.org/sycl/Specification/-/issues/564).

* The statement "The `marray` definition used within this proposal assumes that any operator the `marray` class defines is only implemented if the marray's value type also implements the operator" also seems useful.  This is also missing from SYCL 2020, but we have a Khronos issue to add wording like this (https://gitlab.khronos.org/sycl/Specification/-/issues/559).

Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: > * I agree with @rolandschulz that the line "The `marray` class is extended for the `sycl::ext::oneapi::complex` class." is unneeded.  I do not think we are extending that class.  In fact, the next sentence says that the interface is not changed, so these two statements seems to be contradictory.

This makes sense to me. I have removed this line.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc (hunk 19 above)
Comment (rolandschulz) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:120]: > Applications using arrays of complex values will likely want an interface that performs similar operations to the std::complex interface.

I can see that. The question is why should that need be met by SYCL (initially as extension)?I don't think we want to add all math functionality applications reasonable want to have and slowly end up with an API size similar to libraries like Eigen. To keep the API reasonable small we should only have in SYCL what can be considered fundamental building blocks and/or can't be implemented (efficiently) as library.  And I don't think `make_complex_marray` qualifies as such.

Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc]: So regarding your point, I think that if these functions were made member functions it may be less intrusive and awkward in the spec. I have updated the proposal to specialize `marray` and made these functions members of the class. Let me know your thoughts on this approach and if this addresses some of your concerns.

**Review by keryell — COMMENTED** [review]
I have the feeling that the classical operations could be `constexpr`, at least for integral types.

**Comment by keryell** [comment]
> > I have the feeling that the classical operations could be `constexpr`, at least for integral types.
> 
> Hi, thanks for the comment, could you clarify what classical operations you are referring to, is this the arithmetic operations `+`,`-`, etc or classical complex functions like `real`/`imag`? Additionally what specifically do you mean by integral types, is this when the underlying complex value type is an integer ex: `marray<complex<int>>` or are you referring to the `integral_sequences`?

I was thinking to simple operations like `-`, `+`, etc. But actually whatever we can. :-)
https://en.cppreference.com/w/cpp/concepts/integral
For floating point and complex types, there might be some differences between `constexpr` and runtime computation because the binary representation is not part of the standard.
That said, SYCL makes a lot of assumptions inherited from OpenCL and SPIR about data representations, so perhaps we can be optimistic and make everything `constexpr`?

**Comment by AidanBeltonS** [comment]
> I was thinking to simple operations like `-`, `+`, etc. But actually whatever we can. :-)
> https://en.cppreference.com/w/cpp/concepts/integral
> For floating point and complex types, there might be some differences between `constexpr` and runtime computation because the binary representation is not part of the standard.
> That said, SYCL makes a lot of assumptions inherited from OpenCL and SPIR about data representations, so perhaps we can be optimistic and make everything `constexpr`?

So for operators, it seems bad practice to have the complex specialization of the `marray` type use different keywords to the generic `marray` type. Perhaps these operators should specify `constexpr`. Still, that change and discussion should probably be for the general `marray` class, not the complex extension.

That being said I think `real` and `imag`, just the overloads that return values, should be `constexpr` as this would match the `std::complex` interface. Additionally, I think it is safe to have this be a general change, not just for the integral type. I have updated this proposal to reflect the keyword change.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 21]
```diff
@@ -0,0 +1,475 @@
...
+
+            template <std::size_t... I>
+            constexpr marray(const marray &cmplx, std::integer_sequence<std::size_t, I...> int_seq);
+
+            // Available only when: NumElements == 1
+            operator DataT() const;
+
+            static constexpr std::size_t size() noexcept;
+
+            // real and imag
```
Comment (jle-quel) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:154]: @TApplencourt The problem that we were talking about with the ambiguous getter and setters, as well as the free functions, has already been discussed here.

**Comment by TApplencourt** [comment]
Sorry for chiming in for the `constructor`, `getter`, `setter` discussion. IMO for this work, we should be supper boring and follow `std::complex`. So I concur with @rolandschulz 

- The getter should be named `real` and `imag` so I can do `std::marray<complex<T> A ; A.real()`
- The setter should just reuse `&=` with the same semantic as `std::complex`. If I pass a `marray<T>` it will set only the real part, the imaginary part will be `0`.
- The constructor should mimic the std::complex (take only marray as arguments). No need for free function.

If people want to set/get an individual element of the marray, they can just use the `[]` operator and wrote their own wrapper. Lets keep this API simple.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc [hunk 22]
```diff
@@ -0,0 +1,475 @@
...
+
+            // Additional constructors
+            constexpr marray(const marray<T, NumElements> &real, const marray<T, NumElements> &imag);
+
+            constexpr marray(const marray<T, NumElements> &real, T imag);
+
+            constexpr marray(T real, const marray<T, NumElements> &imag);
+
+            template <std::size_t... I>
+            constexpr marray(const marray<T, NumElements> &real, const marray<T, NumElements> &imag, std::integer_sequence<std::size_t, I...> int_seq);
```
[the remark this answers is not in the document]
Comment (jle-quel) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc:144]: Apart from this marray's [test](https://github.com/argonne-lcf/SyclCPLX/blob/3c05fda15f127b658990ddf5d43602bb54834b12/tests/test_marray_free_functions.cpp), I'm not aware of others using these constructors and getters.

**Comment by jle-quel** [comment]
Should we organize a meeting where we could talk about our different opinions, and conclude on what we should do?

Otherwise, do we agree with @TApplencourt and @rolandschulz, plus the removal of the constructors and getters using ```std::integer_sequence```as @keryell mentioned?

**Comment by TApplencourt** [comment]
I'm open to having any other meetings if that help moving forward.

**Comment by TApplencourt** [comment]
Should we vote to move this issue forward?  
I'm not sure democracy is the best solution to solve technical issues, but it should make it moving. 

0️⃣ Minimal API ( removal of most of the getter, setter, and integral sequence, ...)
1️⃣ Keep it as is (integral sequence, set_real, ...)

Please use "emoji reaction" to vote.
And Happy new year :)

**Comment by TApplencourt** [comment]
The people voted! Can we modify the PR to remove the API and you the "light" version? Thanks!

**Comment by jle-quel** [comment]
Due to the new PR introducing the restructured marray specialization proposal, I will close this PR.

Please refer to the new PR https://github.com/intel/llvm/pull/8852
